src/main/java/org/takes/rq/RqHref.java

```diff
@@ -24,50 +24,78 @@
 package org.takes.rq;
 
 import java.io.IOException;
+import java.io.InputStream;
 import java.util.Iterator;
 import lombok.EqualsAndHashCode;
 import org.takes.Request;
 import org.takes.misc.Href;
 
 /**
- * Request decorator, for HTTP URI query parsing.
+ * HTTP URI query parsing.
  *
- * <p>The class is immutable and thread-safe.
+ * <p>All implementations of this interface must be immutable and thread-safe.
  *
  * @author Yegor Bugayenko (yegor@teamed.io)
  * @version $Id$
  * @since 0.9
  */
-@EqualsAndHashCode(callSuper = true)
-public final class RqHref extends RqWrap {
-
-    /**
-     * Ctor.
-     * @param req Original request
-     */
-    public RqHref(final Request req) {
-        super(req);
-    }
-
+public interface RqHref extends Request {
     /**
      * Get HREF.
      * @return HTTP href
      * @throws IOException If fails
      */
-    public Href href() throws IOException {
-        final Iterator<String> host = new RqHeaders(this)
-            .header("Host").iterator();
-        if (!host.hasNext()) {
-            throw new IOException("Host header is absent");
+    Href href() throws IOException;
+
+    /**
+     * Request decorator, for HTTP URI query parsing.
+     *
+     * <p>The class is immutable and thread-safe.
+     * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
+     * @version $Id$
+     * @since 0.13.1
+     */
+    @EqualsAndHashCode(of = "origin")
+    final class Base implements RqHref {
```

Comment:
@alevohin the modifier public is redundant for inner classes in interfaces, isn't it ?